nexus/src/db/datastore.rs

```diff
@@ -1321,6 +1323,83 @@ impl DataStore {
         Ok(())
     }
 
+    /// Return the information about an instance's network interfaces required
+    /// for the sled agent to instantiate it via OPTE.
+    ///
+    /// OPTE requires information that's currently split across the network
+    /// interface and VPC subnet tables. This query just joins those for each
+    /// NIC in the given instance.
+    pub(crate) async fn derive_guest_network_interface_info(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+    ) -> ListResultVec<sled_client_types::NetworkInterface> {
+        opctx.authorize(authz::Action::ListChildren, authz_instance).await?;
+
+        use db::schema::network_interface;
+        use db::schema::vpc_subnet;
+
+        // The record type for the results of the below JOIN query
+        #[derive(Debug, diesel::Queryable)]
+        struct NicInfo {
+            name: db::model::Name,
+            ip: ipnetwork::IpNetwork,
+            mac: db::model::MacAddr,
+            ipv4_block: db::model::Ipv4Net,
+            ipv6_block: db::model::Ipv6Net,
+            vpc_id: Uuid,
+            slot: i16,
+        }
+
+        impl From<NicInfo> for sled_client_types::NetworkInterface {
+            fn from(nic: NicInfo) -> sled_client_types::NetworkInterface {
+                let ip_subnet = if nic.ip.is_ipv4() {
+                    external::IpNet::V4(nic.ipv4_block.0)
+                } else {
+                    external::IpNet::V6(nic.ipv6_block.0)
+                };
+                sled_client_types::NetworkInterface {
+                    name: sled_client_types::Name::from(&nic.name.0),
+                    ip: nic.ip.ip().to_string(),
+                    mac: sled_client_types::MacAddr::from(nic.mac.0),
+                    subnet: sled_client_types::IpNet::from(ip_subnet),
+                    vpc_id: nic.vpc_id,
+                    slot: u8::try_from(nic.slot).unwrap(),
+                }
+            }
+        }
+
+        let rows = network_interface::table
+            .filter(network_interface::instance_id.eq(authz_instance.id()))
+            .filter(network_interface::time_deleted.is_null())
+            .inner_join(
+                vpc_subnet::table
+                    .on(network_interface::subnet_id.eq(vpc_subnet::id)),
+            )
+            .order_by(network_interface::slot)
+            // TODO-cleanup: Having to specify each column again is less than
+            // ideal. However, this type doesn't appear to have the `Row`
+            // associated type in the documentation. DRY this out.
+            .select((
+                network_interface::name,
+                network_interface::ip,
+                network_interface::mac,
+                vpc_subnet::ipv4_block,
+                vpc_subnet::ipv6_block,
+                network_interface::vpc_id,
+                network_interface::slot,
+            ))
```

Comment:
I wanted to include the explain output here, for future reference. The query is using indexes and lookup joins, so it seems it will probably scale well.

```
root@[fd00:1122:3344:1::2]:32221/omicron> explain analyze select network_interface.name, network_interface.ip, network_interface.mac, vpc_subnet.ipv4_block, vpc_subnet.ipv6_block, vpc.vni, network_interface.slot FROM ((network_interface
INNE
))
IS NULL)
                                                     info
----------------------------------------------------------------------------------------------------------------
 planning time: 2ms
 execution time: 2ms
 distribution: full
 vectorized: true
 cumulative time spent in KV: 507µs
 maximum memory usage: 40 KiB
 network usage: 0 B (0 messages)

 • sort
 │ nodes: n1
 │ actual row count: 0
 │ estimated row count: 1
 │ order: +slot
 │
 └── • lookup join
     │ nodes: n1
     │ actual row count: 0
     │ KV rows read: 0
     │ KV bytes read: 0 B
     │ estimated row count: 1
     │ table: vpc@primary
     │ equality: (vpc_id) = (id)
     │ equality cols are key
     │
     └── • lookup join
         │ nodes: n1
         │ actual row count: 0
         │ KV rows read: 0
         │ KV bytes read: 0 B
         │ estimated row count: 1
         │ table: vpc_subnet@primary
         │ equality: (subnet_id) = (id)
         │ equality cols are key
         │
         └── • index join
             │ nodes: n1
             │ actual row count: 0
             │ KV rows read: 0
             │ KV bytes read: 0 B
             │ estimated row count: 1
             │ table: network_interface@primary
             │
             └── • scan
                   nodes: n1
                   actual row count: 0
                   KV rows read: 0
                   KV bytes read: 0 B
                   estimated row count: 1 (100% of the table; stats collected 3 hours ago)
                   table: network_interface@network_interface_instance_id_name_key (partial index)
                   spans: [/'2fad36b0-312a-498f-84b0-897f9e762a17' - /'2fad36b0-312a-498f-84b0-897f9e762a17']
(50 rows)

Time: 6ms total (execution 5ms / network 1ms)

root@[fd00:1122:3344:1::2]:32221/omicron>
```

Reply:
Ugh, the actual query is mangled a bit. I transcribed that from the `diesel-dtrace` output, so it should be exactly what's sent to the DB.